flowcontrol/bbr/state.go

```diff
@@ -0,0 +1,180 @@
+package bbr
+
+// This file contains implementations of the various "states" the limiter
+// can be in. This mostly corresponds to the paper's appendix.
+
+import (
+	"math"
+	"math/rand"
+	"time"
+)
+
+type state interface {
+	initialize(lim *Limiter)
+	postAck(lim *Limiter, pm packetMeta, now time.Time)
+	snapshot() state
+}
+
+type startupState struct {
+	// Our bandwidth estimate as of the last time we received an ack:
+	prevBtlBwEstimate bytesPerNs
+
+	// How many times in a row we've tried increasing sending rate and
+	// not seen a signficant bandwidth increase:
+	plateauRounds int
+}
+
+func (s *startupState) initialize(lim *Limiter) {
+	lim.cwndGain = 2 / math.Ln2
```

Comment:
```
// cwndGain (Congestion WiNDow; standard terminology) is a factor
// by which we may exceed BDP. i.e. if inflight() is greater than
// estimated BDP * cwndGain, attempts to send will block.
```


Sanity check:  we do not exceed BDB unless inflight > ~2.89 ?

(Under what conditions is it necessary to exceed BDP again?)

Reply:
In order to test whether our current estimate is too low, we have to try sending faster than that estimate would recommend, to see if we get better bandwidth that way (if so, the estimate was too low, if not, it wasn't). So cwndGain gives us some headroom to do those experiments.

Not sure what your first question is asking? `inflight()` returns the number of bytes currently on the wire (an integer).

Reply:
Makes sense 👍
First question was just verifying that the threshold seemed reasonable.  I forget why, but it caused me some confusion this morning.

Reply:
Yeah, it's about 2.89; the idea is we want it to increase exponentially with each round trip. The paper claims this is the minimum value that will facilitate that, but tbh I don't understand why 2 doesn't work.

Should I go ahead and merge this then, and we'll deal with the remaining issues as a follow up?

Reply:
👍
And yes, go for it!